jffs2/phoenix-rtos/mtd.c

```diff
@@ -265,39 +263,53 @@ int mtd_unpoint(struct mtd_info *mtd, loff_t from, size_t len)
 int mtd_erase(struct mtd_info *mtd, struct erase_info *instr)
 {
 	int ret;
+	uint32_t addr = mtd->start + (instr->addr / mtd->writesize);
+	uint32_t blockno = mtd->start + (instr->addr / mtd->erasesize);
 
 	if (instr->len != mtd->erasesize || instr->addr % mtd->erasesize) {
 		printf("mtd_erase: Invalid flash erase parametrs\n");
 		return -1;
 	}
 
+	/* simulate non-blocking erase (TODO: make it actually non-blocking?) */
+	instr->state = MTD_ERASING;
+
 	mutexLock(mtd->lock);
-	if ((ret = flashdrv_erase(mtd->dma, (uint32_t)(instr->addr / mtd->writesize) + mtd->start))) {
-		printf("mtd_erase: Flash erase error 0x%d\n", ret);
-		return -1;
-	}
+	do {
+		if (flashdrv_isbad(mtd->dma, addr)) {
+			/* this should never happen, jffs2 maintains it's own badblock list */
+			printf("mtd_erase: tried to erase badblock %u, aborting\n", blockno);
+			ret = -1;
+			break;
+		}
+		if ((ret = flashdrv_erase(mtd->dma, addr))) {
+			/* will be marked as a badblock by jffs logics eventually */
+			instr->fail_addr = instr->addr;
+			instr->state = MTD_ERASE_FAILED;
+		}
+		else {
+			instr->state = MTD_ERASE_DONE;
+		}
```

Comment:
Maybe just `if (!flashdrv_isbad(mtd->dma, addr))`? You wouldn't need do..while(0) this way

Reply:
the intent was to remove the `flashdrv_isbad` eventually after ensuring jffs never does that. I see that `instr->state` also isn't set in this case, so I'll fix it in a second

Reply:
@agkaminski can You approve the PR now?